Implement extended URI parser

howjmay opened the pull request:
Implement URI parser which is able to parse IPv6 URI like
tcp://[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:4242 .
IPv6 URI is only accepted if it comes with square brackets and
the port should come after colon.

Only limited types of schemes can be accepted by URI parser.
The possible scheme types are store in a string array.

The current URI parser can accept URL like
"http://iota.tangle.ncku:4242".

howjmay committed:
Implement URI parser for IPv6

Implement URI parser which is able to parse IPv6 URI like
tcp://[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:4242

howjmay committed:
Add newline at end of BUILD

thibault-martinez commented on common/network/tests/test_uri_parser.c:
```diff
@@ -0,0 +1,50 @@
+#include <stdio.h>
```
Insert copyright

thibault-martinez commented on common/network/BUILD:
```diff
@@ -17,3 +17,12 @@ cc_library(
+        ":uri_parser",
+        "@unity",
+    ],
+) 
```
Put this in a separate `BUILD` file in the `tests` folder

thibault-martinez commented on common/network/tests/test_uri_parser.c:
```diff
@@ -0,0 +1,50 @@
+
+  scheme = (char *)malloc(sizeof(char) * scheme_len);
+  host = (char *)malloc(sizeof(char) * host_len);
+  port = (uint16_t *)malloc(sizeof(uint16_t));
```
Just use static variables, no need for dynamic allocation here (which you are not freeing btw)

thibault-martinez commented on common/network/tests/test_uri_parser.c:
```diff
@@ -0,0 +1,50 @@
+
+  scheme = (char *)malloc(sizeof(char) * scheme_len);
+  host = (char *)malloc(sizeof(char) * host_len);
+  port = (uint16_t *)malloc(sizeof(uint16_t));
```
Ditto

thibault-martinez commented on common/network/tests/test_uri_parser.c:
```diff
@@ -0,0 +1,50 @@
+
+  RUN_TEST(test_uri_parser_ipv4);
+  RUN_TEST(test_uri_parser_ipv6);
+
```
When you're at it, maybe test cases where the parser is supposed to fail ?

howjmay replied:
Sorry I don't understand what you mean, since both the test cases showed pass.
Does it would go wrong if I used multiple RUNTEST() ?

thibault-martinez replied:
You are currently testing cases where the parser returns true.
For a complete coverage, you should also add tests where the parser returns false (e.g. bad scheme, missing ':' or ']' etc...)

howjmay replied:
got it. Thank you
I am going to implement it now.

thibault-martinez commented on common/network/uri_parser.c:
```diff
@@ -32,24 +32,45 @@ bool uri_parse(char const *const uri, char *const scheme,
+    }
+    return true;
+  } else {
     return false;
```
No need for an `else`, just `return false`, otherwise you don't have a return for the function

thibault-martinez commented on common/network/uri_parser.c:
```diff
@@ -32,24 +32,45 @@ bool uri_parse(char const *const uri, char *const scheme,
   }
   offset += length;
-  if (strncmp(uri + offset, "://", 3) != 0) {
+  if (strncmp(uri + offset, "://[", 4) == 0) {
```
There is a lot of duplicated code here, could you try to factorize ?

thibault-martinez commented:
Hi @HowJMay , thanks for the PR !

howjmay committed:
Couple fixes

howjmay commented:
Thank you for your advices too

thibault-martinez commented on common/network/uri_parser.c:
```diff
@@ -16,6 +16,10 @@ bool uri_parse(char const *const uri, char *const scheme,
+  enum URI_CASE {
+    ipv4,
+    ipv6
+  } uri_case;
```
Could you put this in `common/network/network.h` ?
```
  typedef enum ip_version_e {
    IPV4,
    IPV6
  } ip_version_t;
```

howjmay replied:
ok

howjmay committed:
Add wrong URI and http URI testing

IPv6 URI is only accepted if it comes with square brackets and
the port should come after colon.

Only limited types of schemes can be accepted by URI parser.
The possible scheme types are store in a string array.

The current URI parser can accept URL like
"http://iota.tangle.ncku:4242".

howjmay committed:
Fix scheme_len and host_len checking

The old if statement to determine whether the parsed
scheme(host) length longer than the scheme(host) length
we assigned as function argument has little problem.

The old length would point to the address that colon locates
which is exactly the same as the string length of scheme(host).

e.g. if we parse "https://.."
(delim = strchr(uri, ':'));
length = (size_t)(delim - (uri + offset));
the above code will return length == 5 which is exactly the
string length of "https".

Therefore, the old one cause the wrong judgement.

howjmay committed:
Fix typo

jserv commented on common/network/network.h:
```diff
@@ -12,5 +12,5 @@
 typedef enum protocol_type_e { PROTOCOL_TCP, PROTOCOL_UDP } protocol_type_t;
-
-#endif  // __COMMON_NETWORK_H__
+typedef enum ip_version_e { IPV4, IPV6 } ip_version_t;
```
Insert a blank line for consistency.

jserv commented on common/network/tests/test_uri_parser.c:
```diff
@@ -0,0 +1,92 @@
+#include <unity/unity.h>
+
+void test_uri_parser_ipv4(void) {
+  char *uri_ipv4 = "tcp://8.8.8.8:4242";
```
If you define macros such as `PROTOCOL`, `IP_ADDR` and `PORT` in advance, the URI would be `PROTOCOL "://" IP_ADDR ":" PORT` while the macros can be used later.

jserv commented on common/network/tests/test_uri_parser.c:
```diff
@@ -0,0 +1,92 @@
+}
+
+void test_uri_parser_http(void) {
+  char *uri_http = "http://iota.tangle.ncku:4242";
```
Avoid listing unpopular host(s) in test items.

jserv commented on common/network/uri_parser.c:
```diff
@@ -24,32 +28,77 @@ bool uri_parse(char const *const uri, char *const scheme,
+  // determine if scheme is acceptable type
+  for (int i = 0; i < possible_scheme_num; i++) {
+    int len;
+    if (strlen(possible_scheme[i]) > strlen(scheme)) {
```
Avoid invoking `strlen` more than twice.

thibault-martinez commented:
Hi, I was fine with the previous version: I prefer to consider this as a field extractor rather than a full URI validator (i.e allowed scheme or counting ':')

howjmay commented:
@thibault-martinez  got it. I will reset it in the next version

howjmay committed:
Define macros and fix couple problems

Protocols, hosts and port are replaced with macros.

'strlen()' are removed from uri_parser.c

unpopular hosts are replaced with the node listed on
"https://iota.dance"

field extrator are restored into the original one

jserv commented on common/network/uri_parser.c:
```diff
@@ -31,25 +36,64 @@ bool uri_parse(char const *const uri, char *const scheme,
     return false;
   }
-  if (host) {
+  if (host != NULL) {
```
Minimize required changes. Don't perform non-functioned line.

jserv commented on common/network/network.h:
```diff
@@ -12,5 +12,6 @@
+typedef enum ip_version_e { IPV4, IPV6 } ip_version_t;
 
-#endif  // __COMMON_NETWORK_H__
+#endif // __COMMON_NETWORK_H__
```
Don't do that. Minimize the changes.

jserv commented on common/network/uri_parser.h:
```diff
@@ -35,4 +44,4 @@ bool uri_parse(char const *const uri, char *const scheme, size_t scheme_len,
 #endif
 
-#endif  // __COMMON_NETWORK_URI_PARSER_H__
+#endif // __COMMON_NETWORK_URI_PARSER_H__
```
Ditto.

jserv reviewed (changes requested):
Minimize the changes.

howjmay committed:
Minimize required changes

howjmay commented:
@thibault-martinez  Hi!
It has been a long time since I sent this PR.
I wonder the status of code review of this PR.
I really hope that my PR can be merged into this porject. Thanks !

thibault-martinez commented:
Hi @HowJMay, I was waiting for us to have a license. I will resume reviewing your 3 PRs today.

howjmay commented:
Thank you so much:)

thibault-martinez commented on common/network/uri_parser.c:
```diff
@@ -16,6 +16,11 @@ bool uri_parse(char const *const uri, char *const scheme,
   size_t offset = 0;
   char *delim = NULL;
+  ip_version_t uri_case;
+  const char *possible_scheme[] = {PROTOCOL_TCP, PROTOCOL_UDP, PROTOCOL_HTTP,
```
IMO the parser should only check for syntactic errors and fill fields but not validate the values of the tokens, please remove this part.

howjmay committed:
add brackets

howjmay committed:
Merge branch 'develop' of github.com:iotaledger/entangled into develop

jserv commented:
Rebasing is required.

howjmay committed:
Merge branch 'develop' of github.com:iotaledger/entangled into develop

howjmay committed:
remove token validation

thibault-martinez commented:
Won't be merged due to ciri being discontinued and entangled soon being archived.
Thanks anyway!
